### PR TITLE
TST: add tests for ?gtsvx

### PR DESCRIPTION
**Pull request by swallan**
#### Reference issue

Adds tests for: Add wrapper for ?gtsvx #11373

#### What does this implement/fix?

This pull request adds tests for the ?gtsvx LAPACK routine.

### CODE REVIEW

**Comment by swallan** [comment]
@mdhaber Here's the pull for my ?gtsvx branch. I think it should be pretty close. (edit: I had time so I added the transposal options tonight)

In addition, since the wrapper for this routine is written we can see the types of the output. `rcond`'s output type looks to be a float. I can change this test to look for that, and I assume that rcond will be then also be a float with ?ptsvx, so I can change that there, too.

Something else to consider is that to fully run this test, we'd need to use the ?gttrf routine, but it's not yet on this branch. Let me know what I should do about that!

**Comment by mdhaber** [comment]
> Something else to consider is that to fully run this test, we'd need to use the ?gttrf routine, but it's not yet on this branch.

In the meantime, you can create a new branch from this one locally, merge Kai's branch with ?gttrf into that, and try running it. However, @Kai-Striega should probably wait until ?gttrf gets merged into master (and rebased, if need be) before opening a PR for this.

File: scipy/linalg/tests/test_lapack.py
Comment (mdhaber) [inline scipy/linalg/tests/test_lapack.py]: I don't see how this will work.  This lambda function returns the wrapper to gttrf, but I think you mean _execute_ gttrf on the inputs and return the results.

File: scipy/linalg/tests/test_lapack.py
Comment (mdhaber) [inline scipy/linalg/tests/test_lapack.py]:
```suggestion
    transposal options, incompatible shapes, singular matrices, and
```

File: scipy/linalg/tests/test_lapack.py
Comment (mdhaber) [inline scipy/linalg/tests/test_lapack.py]:
```suggestion
```

File: scipy/linalg/tests/test_lapack.py
Comment (mdhaber) [inline scipy/linalg/tests/test_lapack.py]:
```suggestion
    # Generate random tridiagonal matrix A
```

File: scipy/linalg/tests/test_lapack.py
Comment (mdhaber) [inline scipy/linalg/tests/test_lapack.py]:
```suggestion
```

File: scipy/linalg/tests/test_lapack.py
Comment (mdhaber) [reply scipy/linalg/tests/test_lapack.py]: Comments are good, but arguably some lines are simple enough that they should be lumped together.

File: scipy/linalg/tests/test_lapack.py
Comment (mdhaber) [inline scipy/linalg/tests/test_lapack.py:1778]: On the other hand, for a line like _this_ a comment would be helpful since you don't use `inputs_cpy` immediately.

File: scipy/linalg/tests/test_lapack.py
Comment (mdhaber) [inline scipy/linalg/tests/test_lapack.py]:
```suggestion
```

File: scipy/linalg/tests/test_lapack.py
Comment (mdhaber) [inline scipy/linalg/tests/test_lapack.py]: Maybe move the LU decomposition tests down here, too, right?

Comment (mdhaber) [inline scipy/linalg/tests/test_lapack.py]:
```suggestion
    # no need to test with fact "F" since ?gttrf already tests for these.
```

**Review by mdhaber — COMMENTED** [review]
Looks pretty good.

File: scipy/linalg/tests/test_lapack.py
Comment (swallan) [reply scipy/linalg/tests/test_lapack.py]: Yes, the `(dl, d, du)` is missing. I will add that in.

File: scipy/linalg/tests/test_lapack.py
Comment (swallan) [reply scipy/linalg/tests/test_lapack.py]: I probably got a little overzealous for some of these!

File: scipy/linalg/tests/test_lapack.py
Comment (swallan) [reply scipy/linalg/tests/test_lapack.py]: Do you mean by the transposal options? I don't think we recreate A with the output when the input matrix is singular since it doesn't complete the factorization. (unless info == N, but it shouldn't be)

File: scipy/linalg/tests/test_lapack.py
Comment (mdhaber) [reply scipy/linalg/tests/test_lapack.py]: Probably better to have too many than too few.

File: scipy/linalg/tests/test_lapack.py
Comment (mdhaber) [reply scipy/linalg/tests/test_lapack.py]: I mean the tests that check that the factors represent an LU decomposition. You only need to check that when ?gtsvx does the factoring, not when the factors are provided.

File: scipy/linalg/tests/test_lapack.py
Comment (mdhaber) [inline scipy/linalg/tests/test_lapack.py]:
```suggestion
                                                                   dtype=d.dtype)[0](dl, d, du)),
```
Doesn't it return a list?

File: scipy/linalg/tests/test_lapack.py
Comment (mdhaber) [inline scipy/linalg/tests/test_lapack.py:1762]:
```suggestion
    gtsvx = get_lapack_funcs('gtsvx', dtype=dtype)[0]
```
I think it returns a list.

File: scipy/linalg/tests/test_lapack.py
Comment (mdhaber) [inline scipy/linalg/tests/test_lapack.py]:
```suggestion
    # store a copy of the inputs so we can check they haven't been modified later
```

File: scipy/linalg/tests/test_lapack.py
Comment (mdhaber) [inline scipy/linalg/tests/test_lapack.py]: ?gttrf returns info, right? You need six outputs.

File: scipy/linalg/tests/test_lapack.py
Comment (mdhaber) [inline scipy/linalg/tests/test_lapack.py]: Only needed with `fact='N'`

File: scipy/linalg/tests/test_lapack.py
Comment (mdhaber) [inline scipy/linalg/tests/test_lapack.py:1807]: Check for `info=0`
Please review this against the original issue gh-11373.

File: scipy/linalg/tests/test_lapack.py
Comment (mdhaber) [inline scipy/linalg/tests/test_lapack.py]:
```suggestion
        d[-1] = 0
```

File: scipy/linalg/tests/test_lapack.py
Comment (mdhaber) [inline scipy/linalg/tests/test_lapack.py]:
```suggestion
        dl[-2] = 0
```
Below, too.

File: scipy/linalg/tests/test_lapack.py
Comment (mdhaber) [inline scipy/linalg/tests/test_lapack.py]: Should these have underscores? You're modifying `df` and passing `df_` into `gtsvx`. I'm trying to catch things like this, but please trace each of the variables through the code, too.

File: scipy/linalg/tests/test_lapack.py
Comment (mdhaber) [inline scipy/linalg/tests/test_lapack.py]:
```suggestion
    dl, d, du, _ = inputs_cpy
```

File: scipy/linalg/tests/test_lapack.py
Comment (mdhaber) [inline scipy/linalg/tests/test_lapack.py]: Can we parameterize the trans options?

File: scipy/linalg/tests/test_lapack.py
Comment (mdhaber) [inline scipy/linalg/tests/test_lapack.py]: Since we're checking that the factors represent an LU decomposition in `gttrf`, too, please move all this into a function `L, U = gttrf2lu(dlf, df, duf, du2f, ipiv)` (or similar). Use the function above (and here if you can't parameterize transposal option`, and after the `gttrf` is merged, we should modify its test to use the function.

File: scipy/linalg/tests/test_lapack.py
Comment (mdhaber) [inline scipy/linalg/tests/test_lapack.py]: Should probably break the line after `@pytest.mark.parametrize("du,d,dl,b,x",` so there is more space.

**Review by mdhaber — CHANGES_REQUESTED** [review]

File: scipy/linalg/tests/test_lapack.py
Comment (swallan) [reply scipy/linalg/tests/test_lapack.py]: Sorry I missed that.

File: scipy/linalg/tests/test_lapack.py
Comment (swallan) [reply scipy/linalg/tests/test_lapack.py:1762]: It seems to be working for me without that, for me it is returning a fortran object.

I'll have to commit changes tomorrow morning, as I've got to go to my 7 am tomorrow. Sorry that this wasn't fully finished by today.

File: scipy/linalg/tests/test_lapack.py
Comment (swallan) [reply scipy/linalg/tests/test_lapack.py]: Good call.

File: scipy/linalg/tests/test_lapack.py
Comment (swallan) [reply scipy/linalg/tests/test_lapack.py]: I considered that, but didn't since the complex types and real types take in different trans values, so I thought that there would need to be some sort of if statement anyways. I believe we made the same choice in the test for gttrf/s.

File: scipy/linalg/tests/test_lapack.py
Comment (mdhaber) [inline scipy/linalg/tests/test_lapack.py]: @swallan you could parameterize the trans values like:

```suggestion
@pytest.mark.parametrize("dtype,trans", 
                                            list(zip(DTYPES, ["N"]*len(DTYPES))) 
                                            + list(zip(REAL_DTYPES , ["T"]*len(REAL_DTYPES ))) 
                                            + list(zip(COMPLEX_DTYPES , ["C"]*len(COMPLEX_DTYPES ))))
```
or something like that.

Comment (mdhaber) [reply scipy/linalg/tests/test_lapack.py]: If you don't want to run the whole test with each transpose option, you could also do something like:
```suggestion
@pytest.mark.parametrize("dtype, trans", ["T", "T", "C", "C"])
```

Comment (swallan) [reply scipy/linalg/tests/test_lapack.py]: The first one seems to work pretty well, I like it.

**Comment by swallan** [comment]
@mdhaber Interestingly, when passing in a singular factored matrix (fact=F), it is returning info of n+1 to indicate that RCOND is less than machine precision. Is this particular to my specific machine, or how should I deal with that?

**Comment by mdhaber** [comment]
>  Is this particular to my specific machine, or how should I deal with that?

Probably not. The matrix is exactly singular, so something has to go wrong. Check for that in the test.

**Comment by swallan** [comment]
> Check for that in the test

OK. 

I have run this on a branch with gttrf merged into it, and it appears to run smoothly, other than the singular matrix returning info of n+1.  I double checked to see that the determinant was zero, so it definitely is exactly singular. I also plan on coming into office hours tomorrow, BTW.

File: scipy/linalg/tests/test_lapack.py
Comment (Kai-Striega) [reply scipy/linalg/tests/test_lapack.py:1762]: @mdhaber this is fine. `get_lapack_funcs` returns the fortran function directly if `names` is a string:
```python
Python 3.7.4 (default, Aug  9 2019, 18:34:13) [MSC v.1915 64 bit (AMD64)]
Type 'copyright', 'credits' or 'license' for more information
IPython 7.12.0 -- An enhanced Interactive Python. Type '?' for help.

In [1]: import numpy as np
In [2]: from scipy.linalg import get_lapack_funcs
In [3]: get_lapack_funcs('gtsv', dtype=float)
Out[3]: <fortran object>
In [4]: get_lapack_funcs(('gtsv',), dtype=float)
Out[4]: [<fortran object>]
```

**Comment by Kai-Striega** [comment]
@swallan sorry about the delay. I suspect the `n + 1` info is due to (not) converting to/from 1 based indexes used in Fortran.

**Comment by Kai-Striega** [comment]
@mdhaber I see that requested some changes, are you satisfied that they have been implemented? And are there any other changes you would like to see or are you happy for me to merge this?

I am also considering converting the `info` argument to be `n` not `n + 1`, however I think that's a simple enough fix for me to make on my branch.

File: scipy/linalg/tests/test_lapack.py
Comment (mdhaber) [reply scipy/linalg/tests/test_lapack.py:1762]: Ah.

File: scipy/linalg/tests/test_lapack.py
Comment (mdhaber) [inline scipy/linalg/tests/test_lapack.py]: @swallan I saw this assert message in the last PR, too. What does it mean?

I think I understand what you're going for. If this assert fails, then `info` is not correct, because `d[info-1]` is not the problem. But I'm not sure that's the right thing to assert. The documentation states that:
```
If some U(i,i)=0, so that U is exactly singular, then the routine
    returns with INFO = i
```
It doesn't say:
```
If some d(i)=0, so that A is exactly singular, then the routine
    returns with INFO = i
```
I'm not certain, but I don't think it's correct in general to assume that because d(i) = 0 that U(i, i) will be zero. The matrix may be permuted before it is factored. Personally, I would just assert that `info > 0`, but if really want to make the test more specific, I think it would be more appropriate to form U and check than U(i,i) is 0.

File: scipy/linalg/tests/test_lapack.py
Comment (mdhaber) [inline scipy/linalg/tests/test_lapack.py]:
```suggestion
    singular factorizations. It parametrizes DTYPES and the `fact` value along
```
I would do something to indicate that "fact" is the name of a variable, not some jargon that the reader doesn't understand.

File: scipy/linalg/tests/test_lapack.py
Comment (mdhaber) [inline scipy/linalg/tests/test_lapack.py]:
```suggestion
```

File: scipy/linalg/tests/test_lapack.py
Comment (mdhaber) [reply scipy/linalg/tests/test_lapack.py]: I'm glad you agree, but please also implement the change.

File: scipy/linalg/tests/test_lapack.py
Comment (mdhaber) [inline scipy/linalg/tests/test_lapack.py]:
```suggestion
        assert_(0 < info <= n, "incorrect info for singular matrix")
```
Right? We haven't subtracted 1 from `info` yet.

**Review by mdhaber — CHANGES_REQUESTED** [review]
See comments. Also, does this test for incorrect/incompatible array sizes for each array? Maybe should split into separate tests to make that more noticeable and to respond to comments in other PRs about breaking down the megatests.

Also, @Kai-Striega did you add `class TestGtsvx`? 
I think it duplicates what is in this PR.

File: scipy/linalg/tests/test_lapack.py
Comment (swallan) [reply scipy/linalg/tests/test_lapack.py]: > don't think it's correct in general to assume that because d(i) = 0 that U(i, i) will be zero

I see that this might be an issue–I had assumed it and it appeared to be correct for this specific case, but we can just stick with `info > 0` so it would always work.

**Comment by swallan** [comment]
Split, and checked for array sizes. 

@Kai-Striega, Should the error message for incorrectly sized array `b` ValueError? It seemed to raise a `_flapack.error`.

**Comment by mdhaber** [comment]
Thanks for these changes. There were also a few places where I suggested `[n-1]` should be `[-1]` and similar.

**Comment by Kai-Striega** [comment]
@swallan the ``_flapack.error`` is fine, that occurs when a check in the wrapper isn't true. You can catch these by using the more general ``Exception`` rather than a `ValueError`

**Comment by Kai-Striega** [comment]
@mdhaber I did add the tests for gtsvx class. We can leave it in here and I'll remove it once merged.

**Comment by mdhaber** [comment]
>We can leave it in here and I'll remove it once merged.

Sure, sounds good.

File: scipy/linalg/tests/test_lapack.py
Comment (Kai-Striega) [inline scipy/linalg/tests/test_lapack.py:1813]: I'm having trouble trying to understand what you are trying to do here. If I'm correct you want to avoid testing the complex conjugates with real data types. In that case it could be easier to use a list comprehension similar to that I used in ?tbtrs:

```python
@pytest.mark.parametrize('dtype,trans',
    [(dtype, trans)
    for dtype in DTYPES for trans in ['N', 'T', 'C']
    if not (trans == 'C' and dtype in REAL_DTYPES)])
```

File: scipy/linalg/tests/test_lapack.py
Comment (Kai-Striega) [inline scipy/linalg/tests/test_lapack.py]: I also think this could be neatened up a little. Defining it inside the parameter already uses quite a bit of white space, maybe if you define it as a function, as you did with `gttrf_to_lu` above 👍  will help? OTOH some people may not like having the additional top-level functions. @mdhaber any thoughts?

File: scipy/linalg/tests/test_lapack.py
Comment (Kai-Striega) [inline scipy/linalg/tests/test_lapack.py]: This comment tells me *what* the line below is doing. I can already see that from the line below. What I really need to understand from the comment is *why* are we doing that?

File: scipy/linalg/tests/test_lapack.py
Comment (mdhaber) [reply scipy/linalg/tests/test_lapack.py:1813]: This code was my suggestion, actually. It tests "N" with all data types, it tests "T" with the real data types, and it tests "C" with the complex data types.  Should we be testing  "T" with all data types?

File: scipy/linalg/tests/test_lapack.py
Comment (mdhaber) [reply scipy/linalg/tests/test_lapack.py]: Yeah that looks really ugly ; ) Let's define it outside @parametrize.

File: scipy/linalg/tests/test_lapack.py
Comment (swallan) [reply scipy/linalg/tests/test_lapack.py]: The lambda expression was my idea for how to parametrize the `fact` value, so that instead of branching based on` fact == N or T` every time, we could use the lambda to get `dlf_, df_, duf_, du2f_, ipiv_, info_lambda` where when `fact = N`, the lambda returns Nones, and when `fact = T`, lambda returns the output of gttrf.

update: a new function will be used for this instead of parametrization.

File: scipy/linalg/tests/test_lapack.py
Comment (swallan) [reply scipy/linalg/tests/test_lapack.py]: Ok!

File: scipy/linalg/tests/test_lapack.py
Comment (mdhaber) [reply scipy/linalg/tests/test_lapack.py]: @swallan can you respond to @Kai-Striega's suggestion by updating the comment?

**Comment by mdhaber** [comment]
@swallan please do implement these when you get a chance so @Kai-Striega can merge.

**Comment by mdhaber** [comment]
I'm going to merge these two in my own branch and submit a PR.

**Pull request closed without merging**
